src/ripple/app/ledger/Ledger.cpp

```diff
@@ -958,142 +1018,109 @@ Ledger::read (Keylet const& k) const
 
 //------------------------------------------------------------------------------
 
-class Ledger::tx_iterator_impl
-    : public BasicView::iterator_impl
+auto
+Ledger::txsBegin() const ->
+    std::unique_ptr<txs_type::iter_base>
 {
-private:
-    SHAMap::iterator iter_;
-
-public:
-    explicit
-    tx_iterator_impl (SHAMap::iterator iter)
-        : iter_(iter)
-    {
-    }
-
-    std::unique_ptr<iterator_impl>
-    copy() const override
-    {
-        return std::make_unique<
-            tx_iterator_impl>(
-                iter_);
-    }
-
-    bool
-    equal (iterator_impl const& impl) const override
-    {
-        auto const& other = dynamic_cast<
-            tx_iterator_impl const&>(impl);
-        return iter_ == other.iter_;
-    }
-
-    void
-    increment() override
-    {
-        ++iter_;
-    }
+    return std::make_unique<
+        txs_iter_impl>(closed(),
+            txMap_->begin(), *this);
+}
 
-    txs_type::value_type
-    dereference() const override
-    {
-        return deserializeTxPlusMeta(**iter_);
-    }
-};
+auto
+Ledger::txsEnd() const ->
+    std::unique_ptr<txs_type::iter_base>
+{
+    return std::make_unique<
+        txs_iter_impl>(closed(),
+            txMap_->end(), *this);
+}
 
 bool
-Ledger::txEmpty() const
+Ledger::txExists (uint256 const& key) const
 {
-    return txMap_->getHash().isZero();
+    return txMap_->hasItem (key);
 }
 
 auto
-Ledger::txBegin() const ->
-    std::unique_ptr<iterator_impl>
+Ledger::txRead(
+    key_type const& key) const ->
+        tx_type
 {
-    // Can't iterate open Ledger objects
-    // because they don't have the metadata!
-    assert(closed());
-    return std::make_unique<
-        tx_iterator_impl>(txMap_->begin());
+    auto const& item =
+        txMap_->peekItem(key);
+    if (! item)
+        return {};
+    if (closed())
+    {
+        auto result =
+            deserializeTxPlusMeta(*item);
+        return { std::move(result.first),
+            std::move(result.second) };
+    }
+    return { deserializeTx(*item), nullptr };
 }
 
 auto
-Ledger::txEnd() const ->
-    std::unique_ptr<iterator_impl>
+Ledger::digest (key_type const& key) const ->
+    boost::optional<digest_type>
 {
-    // Can't iterate open Ledger objects
-    // because they don't have the metadata!
-    assert(closed());
-    return std::make_unique<
-        tx_iterator_impl>(txMap_->end());
+    digest_type digest;
+    // VFALCO Unfortunately this loads the item
+    //        from the NodeStore needlessly.
+    if (! stateMap_->peekItem(key, digest))
+        return boost::none;
+    return digest;
 }
 
 //------------------------------------------------------------------------------
 
-bool
-Ledger::unchecked_erase(
-    uint256 const& key)
+void
+Ledger::rawErase(std::shared_ptr<
```

Comment:
This idea of breaking _within a classname_ is also really confusing to me, particularly since you only do it about one time in five...